lib/queuePopulator/IngestionReader.js

```diff
@@ -0,0 +1,233 @@
+const arsenal = require('arsenal');
+const async = require('async');
+const LogConsumer = arsenal.storage.metadata.bucketclient.LogConsumer;
+const BucketClient = require('bucketclient').RESTClient;
+const ReplicationQueuePopulator =
+    require('../../extensions/replication/ReplicationQueuePopulator');
+const IngestionProducer = require('./IngestionProducer');
+
+const { metricsExtension, metricsTypeQueued } =
+    require('../../extensions/replication/constants');
+
+const LogReader = require('./LogReader');
+
+class IngestionReader extends LogReader {
+    constructor(params) {
+        const { zkClient, kafkaConfig, bucketdConfig, ingestionPath, qpConfig,
+            raftId, logger, extensions, metricsProducer, s3Config } = params;
+        const { host, port } = bucketdConfig;
+        logger.info('initializing raft log reader',
+            { method: 'RaftLogReader.constructor',
+                bucketdConfig, raftId });
+        const bucketClient = new BucketClient(`${host}:${port}`);
+        const logConsumer = new LogConsumer({ bucketClient,
+            raftSession: raftId,
+            logger });
+        super({ zkClient, kafkaConfig, logConsumer, logId: `raft_${raftId}`,
+                logger, extensions, metricsProducer });
+        this.raftId = raftId;
+        this._iProducer = new IngestionProducer(bucketdConfig, qpConfig,
+            s3Config);
+        this.bucketdPrefix = bucketdConfig.prefix ? bucketdConfig.prefix :
+            bucketdConfig.name;
+        this.newIngestion = false;
+        this.removeLogOffset = null;
+        if (ingestionPath) {
+            this.pathToLogOffset = `${ingestionPath}` +
+                `${bucketdConfig.zookeeperSuffix}` +
+                `/logState/raft_${raftId}/logOffset`;
+        }
+    }
+
+    /* eslint-disable no-param-reassign */
+
+    _processReadRecords(params, batchState, done) {
```

Comment:
As a general comment, overriding private functions to get the desired behavior is usually a poor design choice, as the parent class should not care about side-effects of changing its private implementation that could break children classes. If inheritance is to be used (but usually it can be avoided with other means) it's better to define clear interfaces of what functions the child class is allowed to use and what is it allowed to override.